libbpf-tools/sysinjector.c

```diff
@@ -0,0 +1,175 @@
+"\n"
+"Usage: sysinjector [-h] [-p PID] [-c COMM] [-r RETVAL] SYSCALL\n"
+"\v"
+"Examples:\n"
+"  ./sysinjector -r -12 -p 1111 __x64_sys_mknodat            # return ENOMEM\n"
+"  ./sysinjector -r -22 -c sleep __x64_sys_clock_nanosleep   # return EINVAL\n"
+;
+
+static const struct argp_option opts[] = {
+	{ "pid", 'p', "PID", 0, "Process ID to modify"},
+	{ "comm",  'c', "COMM",  0, "Trace this comm only" },
+	{ "retval", 'r', "RETVAL", 0, "Modified return value" },
+	{ NULL, 'h', NULL, OPTION_HIDDEN, "Show the full help"},
+	{},
+};
+
+static error_t parse_arg(int key, char *arg, struct argp_state *state)
+{
+	struct prog_env *env = state->input;
+	long pid;
+	int len, retval;
+
+	switch (key) {
+	case 'h':
+		argp_state_help(state, stderr, ARGP_HELP_STD_HELP);
+		break;
+	case 'p':
+		errno = 0;
+		pid = strtol(arg, NULL, 10);
+		if (errno || pid <= 0) {
+			warn("Invalid PID: %s\n", arg);
+			argp_usage(state);
+		}
+		env->pid = pid;
+		break;
+	case 'c':
+		env->comm = arg;
+		len = strlen(arg) + 1;
+		env->comm_len = len > TASK_COMM_LEN ? TASK_COMM_LEN : len;
+		break;
+	case 'r':
+		errno = 0;
+		retval = strtol(arg, NULL, 10);
+		if (errno) {
+			warn("Invalid RETVAL: %s\n", arg);
+			argp_usage(state);
+		}
+		env->retval = retval;
+		break;
+	case ARGP_KEY_ARG:
+		if (env->syscall) {
+			warn("Too many syscalls: %s\n", arg);
+			argp_usage(state);
+		}
+		env->syscall = arg;
+		break;
+	case ARGP_KEY_END:
+		if (!env->syscall) {
+			warn("Need the syscall to modify the return value\n");
+			argp_usage(state);
+		}
+		break;
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+	return 0;
+}
+
+static int attach_kprobes(struct sysinjector_bpf *obj)
+{
+	long err;
+
+	obj->links.handle_retval_at_enter = bpf_program__attach_kprobe(obj->progs.handle_retval_at_enter, false, env.syscall);
+	err = libbpf_get_error(obj->links.handle_retval_at_enter);
+	if (err) {
+		warn("failed to attach kprobe: %ld\n", err);
+		return -1;
+	}
+
+	return 0;
+}
+
+static volatile int exiting = 0;
+
+static void sig_hand(int signr)
+{
+	exiting = 1;
+}
+
+static struct sigaction sigact = {.sa_handler = sig_hand};
+
+int main(int argc, char **argv)
+{
+	static const struct argp argp = {
+		.options = opts,
+		.parser = parse_arg,
+		.args_doc = args_doc,
+		.doc = program_doc,
+	};
+	struct sysinjector_bpf *obj;
+	int err;
+
+	err = argp_parse(&argp, argc, argv, 0, NULL, &env);
+	if (err)
+		return err;
+
+	sigaction(SIGINT, &sigact, 0);
+
+	err = bump_memlock_rlimit();
+	if (err) {
+		warn("failed to increase rlimit: %d\n", err);
+		return 1;
+	}
+
+	obj = sysinjector_bpf__open();
+	if (!obj) {
+		warn("failed to open BPF object\n");
+		return 1;
+	}
+
+	obj->rodata->targ_tgid = env.pid;
+	obj->rodata->retval = env.retval;
+	if (env.comm)
+		strncpy((char*)obj->rodata->targ_comm, env.comm, env.comm_len);
+
+	err = sysinjector_bpf__load(obj);
+	if (err) {
+		warn("failed to load BPF object\n");
+		return 1;
+	}
+
+	err = attach_kprobes(obj);
```

Comment:
You can use `bpf_program__set_attach_target` and let libbpf do auto-attach.

Reply:
>You can use bpf_program__set_attach_target and let libbpf do auto-attach.

This tool uses kprobes to override the return value.
I think it is the only way to do it.

Reply:
```c
bpf_program__set_attach_target(obj->progs.handle_retval_at_enter, 0, env.syscall);
sysinjector_bpf__attach();
```

Reply:
bpf_program__set_attach_target() is used for BTF-based kernel attach targets.
As of now, when the user-space program specifies attach_btf_id using bpf_program__set_attach_target(), BPF_PROG_TYPE_TRACING, BPF_PROG_TYPE_LSM , BPF_PROG_TYPE_STRUCT_OPS, BPF_PROG_TYPE_EXT are only supported as the program type.
https://github.com/torvalds/linux/blob/v5.16-rc4/kernel/bpf/syscall.c#L2053-L2075
Actually, the above code could not load the BPF skeleton.

Or do you mean we should support both kprove and fmod_ret?

Reply:
You are right, currently libbpf don't support set attach target for kprobe.